workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/LayeredDialogSessions.kt

```diff
@@ -298,16 +301,36 @@ public class LayeredDialogSessions private constructor(
           "Expected a ViewTreeLifecycleOwner on $view"
         }
       }.also { dialogs ->
+        fun closeAll() {
+          dialogs.update(emptyList(), ViewEnvironment.EMPTY) {}
+        }
 
-        val boundsListener = OnGlobalLayoutListener {
-          if (view.getGlobalVisibleRect(boundsRect) && boundsRect != bounds.value) {
-            bounds.value = Rect(boundsRect)
-          }
-          // Should we close the dialogs if getGlobalVisibleRect returns false?
-          // https://github.com/square/workflow-kotlin/issues/599
+        // We rely on the hosting View's WorkflowLifecycleOwner to tell us to tear things down.
+        // WorkflowLifecycleOwner gets hooked up when the View is attached to its window.
+        // But the Activity might finish before the hosting view is ever attached. And we have
+        // lots of time to show Dialogs before then. They will leak.
+        //
+        // To guard against that we hang a default observer directly off of the Activity that
+        // will close all Dialogs when it is destroyed; and we remove it as soon as the hosting
+        // view is attached for the first time.
+        val failsafe = object : DefaultLifecycleObserver {
+          override fun onDestroy(owner: LifecycleOwner) = closeAll()
+        }
+        lifecycleOwnerFromContext(view.context).lifecycle.addObserver(failsafe)
```

Comment:
Almost want to make this `observeUntilViewAttached` to show how the add and remove go together

Reply:
I don't like to wrap up a few a one-off lines like this, even in the same file. I like to be able to see exactly what's happening.